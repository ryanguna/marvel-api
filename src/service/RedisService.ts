/**
 * Module Dependencies
 */
import Bluebird from 'bluebird';
import crypto from 'crypto';

import config from '@config';
import redis from '@lib/redis';

/**
 * Class to handle all redis operations
 *
 * @default
 * @memberOf RedisService
 */
export default class RedisService {
  protected static MARVEL_BUCKET_KEY = 'marvel';

  protected static ETAG_FIELD = 'etag';

  protected static DATA_FIELD = 'data';

  /**
   * Caches etag with request url as key
   *
   * @static
   * @async
   * @param url string
   * @param etag string
   * @returns {Bluebird<void>}
   * @memberOf RedisService
   */
  public static async setEtag(url: string, etag: string): Bluebird<void> {
    const key = RedisService.buildKey(url);

    await redis.hset(key, RedisService.ETAG_FIELD, etag);
    await redis.expire(key, config.get('REDIS_CACHE_TTL_IN_SECONDS'));
  }

  /**
   * Gets etag from cache
   *
   * @static
   * @async
   * @param url string
   * @returns string
   * @memberOf RedisService
   */
  public static async getEtag(url: string): Bluebird<string> {
    const key = RedisService.buildKey(url);
    const etag = await redis.hget(key, RedisService.ETAG_FIELD);

    return etag || '';
  }

  /**
   * Gets cached response
   *
   * @static
   * @async
   * @param url string
   * @returns string
   * @memberOf RedisService
   */
  public static async getCachedData(url: string): Bluebird<[]> {
    const key = RedisService.buildKey(url);
    const response = await redis.hget(key, RedisService.DATA_FIELD);

    return response ? JSON.parse(response) : [];
  }

  /**
   * Sets cached response
   *
   * @static
   * @async
   * @param url string
   * @param data any
   * @returns string
   * @memberOf RedisService
   */
  public static async setCachedData(url: string, data: any): Bluebird<void> {
    const key = RedisService.buildKey(url);
    const value = JSON.stringify(data);
    const pipeline = redis.pipeline();

    pipeline.hset(key, RedisService.DATA_FIELD, value);
    pipeline.expire(key, config.get('REDIS_CACHE_TTL_IN_SECONDS'));

    await pipeline.exec();
  }

  /**
   * Creates a key from url
   *
   * @static
   * @param url string
   * @returns string
   * @memberOf RedisService
   */
  private static buildKey(url: string): string {
    const hashedUrl = crypto
      .createHash('md5')
      .update(`url:${url}`)
      .digest('hex');

    return `${RedisService.MARVEL_BUCKET_KEY}:${hashedUrl}`;
  }
}
