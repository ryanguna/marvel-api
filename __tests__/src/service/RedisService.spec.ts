import RedisService from 'service/RedisService';

const redis = require('@lib/redis').default; // tslint:disable-line no-var-requires

describe('service/RedisService', () => {
  afterEach(() => {
    redis.hget.mockReset();
    redis.hset.mockReset();
    redis.pipeline.mockReset();
  });

  describe('#getEtag', () => {
    it('should return empty when key is not found', async () => {
      await expect(RedisService.getEtag('/example/url')).resolves.toEqual('');

      expect(redis.hget).toHaveBeenCalledTimes(1);
      expect(redis.hget.mock.calls[0][0]).toMatchSnapshot();
      expect(redis.hget.mock.calls[0][1]).toMatchSnapshot();
    });

    it('should return an etag string if key is found', async () => {
      const expectedEtag = 'f0fbae65eb2f8f28bdeea0a29be8749a4e67acb3';
      redis.hget.mockResolvedValueOnce(expectedEtag);

      await expect(RedisService.getEtag('/example/url')).resolves.toEqual(
        expectedEtag,
      );

      expect(redis.hget).toHaveBeenCalledTimes(1);
      expect(redis.hget.mock.calls[0][0]).toMatchSnapshot();
      expect(redis.hget.mock.calls[0][1]).toMatchSnapshot();
    });
  });

  describe('#getCachedData', () => {
    it('should return empty array when key is not found', async () => {
      await expect(RedisService.getCachedData('/example/url')).resolves.toEqual(
        [],
      );

      expect(redis.hget).toHaveBeenCalledTimes(1);
      expect(redis.hget.mock.calls[0][0]).toMatchSnapshot();
      expect(redis.hget.mock.calls[0][1]).toMatchSnapshot();
    });

    it('should return a parsed json object if key is found', async () => {
      const expectedObject = [
        {
          hello: 'world',
        },
      ];

      await redis.hget.mockResolvedValueOnce(JSON.stringify(expectedObject));

      await expect(RedisService.getCachedData('/example/url')).resolves.toEqual(
        expectedObject,
      );

      expect(redis.hget).toHaveBeenCalledTimes(1);
      expect(redis.hget.mock.calls[0][0]).toMatchSnapshot();
      expect(redis.hget.mock.calls[0][1]).toMatchSnapshot();
    });
  });
});
