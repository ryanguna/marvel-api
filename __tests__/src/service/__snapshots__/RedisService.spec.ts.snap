// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`service/RedisService #getCachedData should return a parsed json object if key is found 1`] = `"marvel:41401ba3d8a73d3873f329f26ba0c0db"`;

exports[`service/RedisService #getCachedData should return a parsed json object if key is found 2`] = `"data"`;

exports[`service/RedisService #getCachedData should return empty array when key is not found 1`] = `"marvel:41401ba3d8a73d3873f329f26ba0c0db"`;

exports[`service/RedisService #getCachedData should return empty array when key is not found 2`] = `"data"`;

exports[`service/RedisService #getEtag should return an etag string if key is found 1`] = `"marvel:41401ba3d8a73d3873f329f26ba0c0db"`;

exports[`service/RedisService #getEtag should return an etag string if key is found 2`] = `"etag"`;

exports[`service/RedisService #getEtag should return empty when key is not found 1`] = `"marvel:41401ba3d8a73d3873f329f26ba0c0db"`;

exports[`service/RedisService #getEtag should return empty when key is not found 2`] = `"etag"`;
